### PR TITLE
ddl: fetch schemas in work pool in case of reporting a error in one goroutine. 

### PR DESCRIPTION
**Pull request by joccau**
### What problem does this PR solve?

Issue Number: Ref https://github.com/pingcap/tidb/issues/54436

Problem Summary:
- When fetchAllSchemasWithTables gets an error from one goroutine, other routines are not canceled.
- Use `errgroup` replace it.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/domain/domain.go [hunk 1]
```diff
@@ -442,16 +443,23 @@ func (do *Domain) fetchAllSchemasWithTables(m *meta.Meta) ([]*model.DBInfo, erro
...
 	}
+	if len(allSchemas) == 0 {
+		return nil, nil
+	}
+
 	splittedSchemas := do.splitForConcurrentFetch(allSchemas)
-	doneCh := make(chan error, len(splittedSchemas))
+
+	eg, ectx := errgroup.WithContext(context.Background())
+	concurrency := mathutil.Min(len(splittedSchemas), 128)
```
Comment (hawkingrei) [inline pkg/domain/domain.go:453]:
```suggestion
	concurrency := min(len(splittedSchemas), 128)
```

Comment (joccau) [reply pkg/domain/domain.go]: Modified

File: pkg/domain/domain.go [hunk 2]
```diff
@@ -481,23 +489,29 @@ func (*Domain) splitForConcurrentFetch(schemas []*model.DBInfo) [][]*model.DBInf
 	return splitted
 }
 
-func (*Domain) fetchSchemasWithTables(schemas []*model.DBInfo, m *meta.Meta, done chan error) {
+func (*Domain) fetchSchemasWithTables(ctx context.Context, schemas []*model.DBInfo, m *meta.Meta) error {
+	failpoint.Inject("failed-fetch-schemas-with-tables", func(_ failpoint.Value) {
```
Comment (lance6716) [inline pkg/domain/domain.go:493]:
```suggestion
	failpoint.Inject("failed-fetch-schemas-with-tables", func() {
```

also change the failpoint argument to `1*return()`

**Review by lance6716 — APPROVED** [review]

File: pkg/domain/domain.go [hunk 3]
```diff
@@ -442,16 +443,23 @@ func (do *Domain) fetchAllSchemasWithTables(m *meta.Meta) ([]*model.DBInfo, erro
...
 		return nil, err
 	}
+	if len(allSchemas) == 0 {
+		return nil, nil
+	}
+
 	splittedSchemas := do.splitForConcurrentFetch(allSchemas)
-	doneCh := make(chan error, len(splittedSchemas))
+
+	eg, ectx := errgroup.WithContext(context.Background())
```
Comment (D3Hunter) [inline pkg/domain/domain.go:452]: use NewErrorGroupWithRecoverWithCtx?

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/domain/domain.go (hunk 3 above)
Comment (joccau) [reply pkg/domain/domain.go:452]: Maybe it isn't need. 
The `WorkerPool` can not run the `ErrorGroupWithRecover`, unless 'WorkerPool` is deleted and use ErrorGroupWithRecover directly.

Comment (lance6716) [reply pkg/domain/domain.go:452]: Why do you want to use `NewWorkerPool`? We can use `SetLimit` on errgroup

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**